enterprise/server/workflow/service/service.go

```diff
@@ -794,6 +798,153 @@ func (ws *workflowService) waitForWorkflowInvocationCreated(ctx context.Context,
 	}
 }
 
+func (ws *workflowService) buildActionHistoryQuery(ctx context.Context, repoUrl string, pattern string, timeLimitMicros int64) (*query_builder.Query, error) {
+	q := query_builder.NewQuery(`SELECT repo_url, pattern, invocation_id, commit_sha, branch_name, created_at_usec, updated_at_usec, duration_usec, invocation_status, success FROM Invocations`)
+	if err := perms.AddPermissionsCheckToQuery(ctx, ws.env, q); err != nil {
+		return nil, err
+	}
+	authenticatedUser, err := ws.env.GetAuthenticator().AuthenticatedUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+	q.AddWhereClause("repo_url = ?", repoUrl)
+	q.AddWhereClause("pattern = ?", pattern)
+	q.AddWhereClause("branch_name IN ?", []string{"master", "main"})
+	q.AddWhereClause("role = ?", "CI_RUNNER")
+	q.AddWhereClause("group_id = ?", authenticatedUser.GetGroupID())
+	q.AddWhereClause("updated_at_usec > ?", timeLimitMicros)
+	q.SetOrderBy("created_at_usec", false)
+	q.SetLimit(30)
+
+	return q, nil
+}
+
+func (ws *workflowService) GetWorkflowHistory(ctx context.Context, req *wfpb.GetWorkflowHistoryRequest) (*wfpb.GetWorkflowHistoryResponse, error) {
+	if ws.env.GetDBHandle() == nil || ws.env.GetOLAPDBHandle() == nil {
+		return nil, status.FailedPreconditionError("database not configured")
+	}
+
+	repos := req.GetRepoUrls()
+	authenticatedUser, err := ws.env.GetAuthenticator().AuthenticatedUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+	// Only fetch workflow data for stuff that has run in the last 7 days.
+	timeLimitMicros := time.Now().Add(-time.Duration(7*24) * time.Hour).UnixMicro()
+
+	// Find the names of all of the actions for each repo.
+	q := query_builder.NewQuery(`SELECT repo_url,pattern,count(1) AS total_runs, countIf(success) AS successful_runs, toInt64(avg(duration_usec)) AS average_duration FROM Invocations`)
+	q.AddWhereClause("repo_url IN ?", repos)
+	q.AddWhereClause("role = ?", "CI_RUNNER")
+	q.AddWhereClause("group_id = ?", authenticatedUser.GetGroupID())
+	q.AddWhereClause("invocation_status = ?", int(inspb.InvocationStatus_COMPLETE_INVOCATION_STATUS))
+	q.AddWhereClause("updated_at_usec > ?", timeLimitMicros)
+	q.SetGroupBy("repo_url,pattern")
+	// XXX: No permissions check here; is this ok?
```

Comment:
The current perms check (group_id = authenticated group ID) LGTM. We don't support the `perms` field in clickhouse currently which means we're effectively using strict group RW perms, which is fine for our current use cases.